File: 2020/python/day10.py
import unittest
import collections


def run(joltages):
    joltages.insert(0, 0)
    joltages.append(max(joltages)+3)
    joltages.sort()
    diff_one = 0
    diff_three = 0
    for i in range(0, len(joltages)-1):
        diff = joltages[i+1] - joltages[i]
        if diff == 3:
            diff_three += 1
        elif diff == 1:
            diff_one += 1
    return diff_one * diff_three


def part1():
    print(run(read_input()))

#     0
#     1
#     4
#     5
#     6 -> 5 | 4
#     7 -> 6 (5|4) | 5 | 4
#    10 -> 7 -> 6 (5|4) | 5 | 4
#    11 -> 7 -> 6 (5|4) | 5 | 4
#    12 -> 11 | 10 ...


def run2(joltages):
    joltages.insert(0, 0)
    joltages.sort()

    all_options = collections.defaultdict(int)
    all_options[0] = 1

    for i in range(1, len(joltages)):
        jolt = joltages[i]
        pv1 = all_options[jolt-1]
        pv2 = all_options[jolt-2]
        pv3 = all_options[jolt-3]
        all_options[jolt] = pv1 + pv2 + pv3

    return all_options[joltages[-1]]


def part2():
    print(run2(read_input()))


def read_input():
    with open('../inputs/input10') as input:
        return [int(line) for line in input if not line.isspace()]


test_input = """16
10
15
5
1
11
7
19
6
12
4"""

test_input2 = """28
33
18
42
31
14
46
20
48
47
24
23
49
45
19
38
39
11
1
32
25
35
8
17
7
9
4
2
34
10
3"""


class TestPart1(unittest.TestCase):

    def test_run(self):
        joltages = [int(l) for l in test_input.split('\n')]
        self.assertEqual(run(joltages), 35)

        joltages = [int(l) for l in test_input2.split('\n')]
        self.assertEqual(run(joltages), 220)


class TestPart2(unittest.TestCase):

    def test_run2(self):
        joltages = [int(l) for l in test_input.split('\n')]
        self.assertEqual(run2(joltages), 8)

        joltages = [int(l) for l in test_input2.split('\n')]
        self.assertEqual(run2(joltages), 19208)


if __name__ == '__main__':
    part1()
    part2()
    unittest.main()

File: 2020/rust-sln/src/day10.rs
use std::collections::HashMap;

fn main() {
    println!("Part1 {}", part1());
    println!("Part2 {}", part2());
}

fn parse(input: &str) -> Vec<isize> {
    input
        .lines()
        .flat_map(|line| line.trim().parse::<isize>())
        .collect()
}

fn read_input() -> &'static str {
    include_str!("../../inputs/input10")
}

fn run1(adapters: &mut Vec<isize>) -> isize {
    adapters.insert(0, 0);
    adapters.push(adapters.iter().max().expect("no entry in adapters") + 3);
    adapters.sort();
    let mut diff_one = 0;
    let mut diff_three = 0;
    for i in 0..(adapters.len() - 1) {
        let diff = adapters[i + 1] - adapters[i];
        if diff == 3 {
            diff_three += 1;
        } else if diff == 1 {
            diff_one += 1;
        }
    }
    diff_one * diff_three
}

fn part1() -> isize {
    run1(&mut parse(read_input()))
}

fn run2(adapters: &mut Vec<isize>) -> isize {
    adapters.insert(0, 0);
    adapters.sort();
    let mut all_options: HashMap<isize, isize> = HashMap::new();
    all_options.insert(0, 1);

    for i in 1..adapters.len() {
        let adapter = adapters[i];
        let pv1 = all_options.get(&(adapter - 1)).unwrap_or(&0).clone();
        let pv2 = all_options.get(&(adapter - 2)).unwrap_or(&0).clone();
        let pv3 = all_options.get(&(adapter - 3)).unwrap_or(&0).clone();
        all_options.insert(adapter, pv1 + pv2 + pv3);
    }
    *all_options
        .get(adapters.last().expect("Adapters is empty"))
        .expect("No option found")
}

fn part2() -> isize {
    run2(&mut parse(read_input()))
}

mod tests {
    use super::*;

    fn test_input() -> &'static str {
        r#"16
10
15
5
1
11
7
19
6
12
4"#
    }
    #[test]
    fn test_parse() {
        let r = vec![16, 10, 15, 5, 1, 11, 7, 19, 6, 12, 4];
        assert_eq!(parse(test_input()), r);
    }

    #[test]
    fn test_run() {
        assert_eq!(run1(&mut parse(test_input())), 35);
    }
}
